### PR TITLE
feat: Add file format resource

### PR DESCRIPTION
**Pull request by wookasz**
## Test Plan
* [ ] acceptance tests
* [ ] unit tests

### CODE REVIEW

**Comment by wookasz** [comment]
This is not my work. This is https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/392 with merge conflicts resolved. Per https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/392#issuecomment-857815429 just fixing up the merge conflict as requested in https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/392#issuecomment-809622416.

I can also make review changes to try to get this pushed through.

**Comment by alldoami** [comment]
Can you run `make docs`?

File: pkg/resources/file_format.go [hunk 1]
```diff
@@ -0,0 +1,1128 @@
...
+	"XML": {
+		"compression",
+		"ignore_utf8_errors",
+		"preserve_space",
+		"strip_outer_element",
+		"disable_snowflake_data",
+		"disable_auto_convert",
+		"skip_byte_order_mark",
+		"trim_space",
+		"null_if",
```
Comment (alldoami) [inline pkg/resources/file_format.go:81]: I don't think there is `trim_space` or `null_if` for XML... looking at this list: https://docs.snowflake.com/en/sql-reference/sql/create-file-format.html#syntax

File: pkg/resources/file_format_acceptance_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: fileFormatConfig(accName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_file_format.test", "name", accName),
+					resource.TestCheckResourceAttr("snowflake_file_format.test", "database", accName),
+					resource.TestCheckResourceAttr("snowflake_file_format.test", "schema", accName),
+					resource.TestCheckResourceAttr("snowflake_file_format.test", "format_type", "CSV"),
+					resource.TestCheckResourceAttr("snowflake_file_format.test", "comment", "Terraform acceptance test"),
+				),
```
Comment (alldoami) [inline pkg/resources/file_format_acceptance_test.go:25]: Just to see if the other parameters work, can you also add a test with other parameters and file types?

File: pkg/resources/file_format.go [hunk 3]
```diff
@@ -0,0 +1,1128 @@
...
+	}
+	data.SetId(dataIDInput)
+
+	return ReadFileFormat(data, meta)
+}
+
+// ReadFileFormat implements schema.ReadFunc
+func ReadFileFormat(data *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	fileFormatID, err := idFromString(data.Id())
```
Comment (alldoami) [inline pkg/resources/file_format.go:534]: Need to create a `fileFormatIDFromString` function.

File: pkg/resources/file_format.go [hunk 4]
```diff
@@ -0,0 +1,1128 @@
...
+	}
+
+	q := builder.Create()
+
+	err := snowflake.Exec(db, q)
+	if err != nil {
+		return errors.Wrapf(err, "error creating file format %v", name)
+	}
+
+	fileFormatID := &schemaScopedID{
```
Comment (alldoami) [inline pkg/resources/file_format.go:517]:
```suggestion
	fileFormatID := &fileFormatID{
```

File: pkg/resources/file_format.go [hunk 5]
```diff
@@ -0,0 +1,1128 @@
...
+		Database: database,
+		Schema:   schema,
+		Name:     name,
```
Comment (alldoami) [inline pkg/resources/file_format.go:520]:
```suggestion
		FileFormatName:     name,
```

**Comment by wookasz** [comment]
I extended the CSV file format acceptance test to cover all possible parameters, will add more but was hoping to get a run-through of the tests first as an easier way to make sure I'm formatting the verification strings correctly

**Comment by wookasz** [comment]
@alldoami added acceptance tests for all format types and fixed the slew of issues that this brought up. Great call on suggesting that.

**Review by alldoami — APPROVED** [review]
Thank you!

**Comment by philippeboyd** [comment]
🙌

**Pull request merged**